### PR TITLE
quick fix API polish

### LINKED ISSUE #167557: outputMatcher.lineMatcher has wrong description
Testing #166762

image

### LINKED ISSUE #167584: Terminal quick fix API: multipleMatches is confusing
Testing #166762

I expected that if this is true and the outputMatcher regex matches multiple times, I would see these matches in the result, but I don't see any difference

### LINKED ISSUE #167586: terminal quick fix contribution: multipleMatches not suggested
Testing #166762

The example in the test item lists `multipleMatches`, which is not proposed. 

image

### LINKED ISSUE #167587: terminal quick fix contribution: warn on invalid properties
Testing #166762

Add a `additionalProperties: false` to the schema to warn on mistyped or invalid properties

image

### LINKED ISSUE #167590: terminal quick fix contribution: improve property descriptions
Testing #166762

image

- lineMatcher: describe that this is can be a regex which will be the input to the provider?
- commandLineMatcher: is this also a pattern?
- anchor: `Which side of the output to anchor the offset and length against` not sure what that means

### LINKED ISSUE #167621: terminal quick fix contribution: list return types
Testing #166762

the provider returns `TerminalQuickFix`. I first didn't see the subtypes `TerminalQuickFixCommandAction` and `TerminalQuickFixOpenerAction`.

- define `TerminalQuickFix` as `type TerminalQuickFix = TerminalQuickFixCommandAction |  TerminalQuickFixOpenerAction`
- align the naming: TerminalQuickFix -> TerminalQuickFixAction

### LINKED ISSUE #167646: Return types from provideTerminalQuickFixes should be concrete
Testing #166762

Extending https://github.com/microsoft/vscode/issues/167621, the return types from the provider should be concrete to match the style of other providers in the API.

### LINKED ISSUE #167649: terminal quick fix contribution: enum members should be numeric and uppercased
Testing #166762

```ts
enum TerminalOutputAnchor {
	top = 'top',
	bottom = 'bottom'
}

enum TerminalQuickFixType {
	command = 'command',
	opener = 'opener'
}
```

Following the style of other VS Code APIs, the enums should be numeric, and their members uppercase. For a random example, the QuickPickItemKind:

```ts
/**
 * The kind of {@link QuickPickItem quick pick item}.
 */
export enum QuickPickItemKind {
	/**
	 * When a {@link QuickPickItem} has a kind of {@link Separator}, the item is just a visual separator and does not represent a real item.
	 * The only property that applies is {@link QuickPickItem.label label }. All other properties on {@link QuickPickItem} will be ignored and have no effect.
	 */
	Separator = -1,
	/**
	 * The default {@link QuickPickItem.kind} is an item that can be selected in the quick pick.
	 */
	Default = 0,
}
```

### LINKED ISSUE #167652: terminal quick fix contribution: providers should return ProviderResults
Testing #166762

"Providers" in VS Code APIs always return `ProviderResults`, we should do the same thing with the new terminal APIs. This does allow the provider to be async, but in some other cases we have aggressive timeouts to encourage providers to be efficient, e.g. the API could ignore results that take longer than 500ms to resolve.

### PR DESCRIPTION
**Pull request by meganrogge**
fix #167649
fix https://github.com/microsoft/vscode/issues/167652
fix https://github.com/microsoft/vscode/issues/167646
fix https://github.com/microsoft/vscode/issues/167621
fix  #167584
fix #167586
fix #167587
fix #167557
fix #167590

Also got rid of a hack where I was using `class` to indicate the `type`

### CODE REVIEW

File: src/vs/platform/terminal/common/xterm/terminalQuickFix.ts [hunk 1]
```diff
@@ -9,6 +9,13 @@ import { CancellationToken } from 'vs/base/common/cancellation';
 import { URI } from 'vs/base/common/uri';
 import { ITerminalCommand } from 'vs/platform/terminal/common/capabilities/capabilities';
 
+export enum TerminalQuickFixType {
+	Command = 0,
+	Opener = 1,
+	Port = 2
+}
+
+
```
Comment (meganrogge) [inline src/vs/platform/terminal/common/xterm/terminalQuickFix.ts:18]:
```suggestion
```

File: src/vs/platform/terminal/common/xterm/terminalQuickFix.ts [hunk 2]
```diff
@@ -50,7 +56,15 @@ export interface ITerminalCommandSelector {
 
 export type TerminalQuickFixActionInternal = IAction | ITerminalQuickFixCommandAction | ITerminalQuickFixOpenerAction;
 export type TerminalQuickFixCallback = (matchResult: ITerminalCommandMatchResult) => TerminalQuickFixActionInternal[] | TerminalQuickFixActionInternal | undefined;
-export type TerminalQuickFixCallbackExtension = (terminalCommand: ITerminalCommand, lines: string[] | undefined, option: ITerminalQuickFixOptions, token: CancellationToken) => Promise<ITerminalQuickFix[] | ITerminalQuickFix | undefined>;
+export type TerminalQuickFixCallbackExtension = (terminalCommand: ITerminalCommand, lines: string[] | undefined, option: ITerminalQuickFixOptions, token: CancellationToken) => Promise<ProviderResult<ITerminalQuickFix[] | ITerminalQuickFix>>;
+
```
Comment (meganrogge) [inline src/vs/platform/terminal/common/xterm/terminalQuickFix.ts:60]: TODO: should this be put somewhere central? I notice other places do something similar

File: src/vs/workbench/api/common/extHost.protocol.ts [hunk 3]
```diff
@@ -70,7 +70,7 @@ import { CandidatePort } from 'vs/workbench/services/remote/common/remoteExplore
 import { ITextQueryBuilderOptions } from 'vs/workbench/services/search/common/queryBuilder';
```
Comment (meganrogge) [inline src/vs/workbench/api/common/extHost.protocol.ts:70]: rm ProviderResult from everywhere except for the .d.ts and thenable is a promise

File: src/vs/workbench/api/browser/mainThreadTerminalService.ts [hunk 4]
```diff
@@ -456,12 +457,14 @@ class ExtensionTerminalLinkProvider implements ITerminalExternalLinkProvider {
...
-	return match ? { regexMatch: match, outputLines: outputMatcher.multipleMatches ? lines : undefined } : undefined;
+	return match ? { regexMatch: match, outputLines: lines } : undefined;
 }
 
 function parseQuickFix(id: string, source: string, fix: TerminalQuickFix): ITerminalQuickFix {
-	if (fix.type === TerminalQuickFixType.opener) {
-		(fix as TerminalQuickFixOpenerAction).uri = URI.revive((fix as TerminalQuickFixOpenerAction).uri);
+	let type = TerminalQuickFixType.Command;
+	if ('uri' in fix) {
+		(fix as TerminalQuickFixOpener).uri = URI.revive((fix as TerminalQuickFixOpener).uri);
```
Comment (Tyriar) [inline src/vs/workbench/api/browser/mainThreadTerminalService.ts:466]: Can you remove the `as`?

File: src/vs/workbench/contrib/terminal/browser/terminalQuickFixBuiltinActions.ts [hunk 5]
```diff
@@ -101,11 +100,13 @@ export function freePort(terminalInstance?: Partial<ITerminalInstance>): IIntern
...
 			const label = localize("terminal.freePort", "Free port {0}", port);
 			return {
-				class: TerminalQuickFixType.Port,
+				type: TerminalQuickFixType.Port,
+				class: undefined,
 				tooltip: label,
 				id: 'Free Port',
 				label,
 				enabled: true,
+				source: 'buitlin',
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/terminalQuickFixBuiltinActions.ts:109]: typo, can that be a const enum?

File: src/vs/workbench/contrib/terminal/common/terminal.ts [hunk 6]
```diff
@@ -752,15 +753,15 @@ export const terminalContributionsDescriptor: IExtensionPointDescriptor<ITermina
...
 							type: 'string',
 						},
 						commandLineMatcher: {
-							description: nls.localize('vscode.extension.contributes.terminal.quickFixes.commandLineMatcher', "The command line to match."),
+							description: nls.localize('vscode.extension.contributes.terminal.quickFixes.commandLineMatcher', "The regular expression to test the command line against."),
 							type: 'string',
 						},
 						outputMatcher: {
-							description: nls.localize('vscode.extension.contributes.terminal.quickFixes.outputMatcher', "The output to match, which provides groups of the form <group_name> to be referenced via ${group:group_name} in commandToRun and linkToOpen."),
+							description: nls.localize('vscode.extension.contributes.terminal.quickFixes.outputMatcher', "The regular expression to test the output against, which provides groups of the form <group_name> to be referenced via ${group:group_name} in terminalCommand and uri."),
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/common/terminal.ts:764]: Do any contributions like this have examples in them? It's hard to understand this without one.

File: src/vs/workbench/contrib/terminal/common/terminal.ts [hunk 7]
```diff
@@ -769,11 +770,11 @@ export const terminalContributionsDescriptor: IExtensionPointDescriptor<ITermina
 									type: 'string'
 								},
 								anchor: {
-									description: 'Which side of the output to anchor the offset and length against',
+									description: 'Where the search should begin in the buffer',
 									enum: ['top', 'bottom']
 								},
 								offset: {
-									description: 'How far from either the top or the bottom of the butter to start matching against.',
+									description: 'The number of lines vertically from the anchor (top or the bottom) of the buffer to start matching against',
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/common/terminal.ts:777]:
```suggestion
									description: 'The number of lines vertically from the anchor in the buffer to start matching against',
```

File: src/vscode-dts/vscode.proposed.terminalQuickFixProvider.d.ts [hunk 8]
```diff
@@ -12,19 +12,16 @@ declare module 'vscode' {
 		 * @param token A cancellation token indicating the result is no longer needed
 		 * @return Terminal quick fix(es) if any
 		 */
-		provideTerminalQuickFixes(commandMatchResult: TerminalCommandMatchResult, token: CancellationToken): TerminalQuickFix[] | TerminalQuickFix | undefined;
+		provideTerminalQuickFixes(commandMatchResult: TerminalCommandMatchResult, token: CancellationToken): ProviderResult<(TerminalQuickFixCommand | TerminalQuickFixOpener)[] | (TerminalQuickFixCommand | TerminalQuickFixOpener)>;
```
Comment (Tyriar) [inline src/vscode-dts/vscode.proposed.terminalQuickFixProvider.d.ts:15]:
```suggestion
		provideTerminalQuickFixes(commandMatchResult: TerminalCommandMatchResult, token: CancellationToken): ProviderResult<(TerminalQuickFixCommand | TerminalQuickFixOpener)[] | TerminalQuickFixCommand | TerminalQuickFixOpener>;
```

File: src/vscode-dts/vscode.proposed.terminalQuickFixProvider.d.ts [hunk 9]
```diff
@@ -36,12 +33,10 @@ declare module 'vscode' {
 		export function registerTerminalQuickFixProvider(id: string, provider: TerminalQuickFixProvider): Disposable;
 	}
 
-	export interface TerminalQuickFixCommandAction extends TerminalQuickFix {
-		type: TerminalQuickFixType.command;
+	export interface TerminalQuickFixCommand {
```
Comment (Tyriar) [inline src/vscode-dts/vscode.proposed.terminalQuickFixProvider.d.ts:36]: Let's make these classes

**Review by Tyriar — CHANGES_REQUESTED** [review]

File: src/vs/workbench/contrib/terminal/common/terminal.ts (hunk 6 above)
Comment (meganrogge) [reply src/vs/workbench/contrib/terminal/common/terminal.ts:764]: we should probably only do that when the API has been finalized though

**Review by Tyriar — APPROVED** [review]
I didn't review again but we talked through this in detail so should be good 👍

**Pull request merged**